### ISSUE TITLE
Compile to WASM

### ISSUE THREAD

**Issue by chpio** [kaj/rsass#33]
https://github.com/sass/node-sass/issues/2011

I've mentioned in that issue, that there could be a self contained "core" library without any external bindings, it would request all the needed data by some kind of abstract async api. That way we could port it more easily to other platforms like WASM (nodejs file io or even the browser).

**Comment by kaj** [comment]
Cool!  I havn't really looked into wasm yet. What would be required to compile rsass into wasm?  Does it have to be no-std?  In that case, I guess some kind of callback-based handling would be needed for supporting `@import`?  Maybe converting `FileContext` to a trait that a caller can provide it's own implementation of?

The one dependency that could probably not be removed from such a "rsass core" would be [nom](https://crates.io/crates/nom).  Is that possible to use in wasm?

**Comment by chpio** [comment]
> Does it have to be no-std?

No, but a some of the std APIs are stubbed out (like [threading and atomics](https://rustwasm.github.io/2018/10/24/multithreading-rust-and-wasm.html)) or panic (filesystem).

> In that case, I guess some kind of callback-based handling would be needed for supporting @import? Maybe converting FileContext to a trait that a caller can provide it's own implementation of?

yeap.

> The one dependency that could probably not be removed from such a "rsass core" would be nom. Is that possible to use in wasm?

I think it should work in wasm. Wasm is a normal [ISA](https://en.wikipedia.org/wiki/Instruction_set_architecture) (like x64/x86), but there are just basic bindings to the host system (like memory allocation), no filesystem and so forth. So as long as it is just using self contained rust code (and allocations) it should work.

**Comment by chpio** [comment]
<https://hacks.mozilla.org/2019/03/standardizing-wasi-a-webassembly-system-interface/>

**Comment by Wicpar** [comment]
@chpio If you do not need runtime modified sass, i made the PR #163 for compile time macros. 
you can try it by adding `rsass-macro = {git = "https://github.com/Wicpar/rsass.git"}` to dependencies and 
```rust
use rsass_macro::scss;
const STR: &'static str = scss!(r#"
  .some-scss {
     // blabla
  }
"#);
```
It works in my wasm dioxus web project.
